csrc/id_model/id_model.cpp

```diff
@@ -0,0 +1,292 @@
+  std::vector<TensorView*> inputs_and_outputs;
+  {
+    auto inp_tvs = ir_utils::filterByType<TensorView>(fusion->inputs());
+    inputs_and_outputs.insert(
+        inputs_and_outputs.begin(), inp_tvs.begin(), inp_tvs.end());
+  }
+  {
+    auto out_tvs = ir_utils::filterByType<TensorView>(fusion->outputs());
+    inputs_and_outputs.insert(
+        inputs_and_outputs.begin(), out_tvs.begin(), out_tvs.end());
+  }
+
+  build(fusion->exprs(), inputs_and_outputs);
+}
+
+const ValGraph& IdModel::idGraph(IdMappingMode mode) const {
+  auto graph_it = id_graphs_.find(mode);
+  NVF_ERROR(graph_it != id_graphs_.end());
+  return graph_it->second;
+}
+
+ValGraph& IdModel::idGraph(IdMappingMode mode) {
+  auto graph_it = id_graphs_.find(mode);
+  NVF_ERROR(graph_it != id_graphs_.end());
+  return graph_it->second;
+}
+
+Expr* IdModel::idUse(IterDomain* id) const {
+  auto use_it = id_uses_.find(id);
+  if (use_it == id_uses_.end()) {
+    return nullptr;
+  }
+  return use_it->second.front();
+}
+
+Expr* IdModel::idDef(IterDomain* id) const {
+  auto def_it = id_definitions_.find(id);
+  if (def_it == id_definitions_.end()) {
+    return nullptr;
+  }
+  return def_it->second.front();
+}
+
+void IdModel::buildIterDomainDefinitionsAndUses(
+    const std::vector<TensorView*>& all_tvs) {
+  for (auto tv : all_tvs) {
+    VectorOfUniqueEntries<IterDomain*> root_domain_ids{
+        tv->getRootDomain().begin(), tv->getRootDomain().end()};
+
+    auto all_ids = ir_utils::allIDsOf(tv);
+
+    // Check is this domain is a consumer of a view-like operation
+    bool view_like_domain = tv->domain()->hasViewLikeRFactor();
+
+    for (auto id : all_ids) {
+      // Check if this id is a view like rfactor id
+      if (view_like_domain && id->isRFactorProduct()) {
+        // If the tensor domain is a view like domain, and the iteration
+        // domain is marked as an rfactor product and is in the rfactor
+        // domain, it's a view like rfactor iteration domain
+        const auto& rfactor_domain = tv->domain()->maybeRFactor();
+        if (std::find(rfactor_domain.begin(), rfactor_domain.end(), id) !=
+            rfactor_domain.end()) {
+          view_rfactor_ids_.emplace(id);
+        }
+      }
+
+      if (id_definitions_.find(id) == id_definitions_.end()) {
+        id_definitions_.emplace(id, VectorOfUniqueEntries<Expr*>{});
+      }
+
+      if (id_uses_.find(id) == id_uses_.end()) {
+        id_uses_.emplace(id, VectorOfUniqueEntries<Expr*>{});
+      }
+
+      auto def = id->definition();
+
+      if (def == nullptr || root_domain_ids.has(id)) {
+        continue;
+      }
+
+      id_definitions_[id].pushBack(def);
+
+      auto inp_ids = ir_utils::filterByType<IterDomain>(def->inputs());
+      for (auto inp_id : inp_ids) {
+        id_uses_[inp_id].pushBack(def);
+      }
+    }
+  }
+}
+
+std::string IdModel::toString() const {
+  // Figure out which graphs are already initialized to make sure we add the new
+  // expression to them.
```

Comment:
> add the new expression

Does this function add any new expressions? Is this an outdated comment copied from somewhere else?

Reply:
I suppose so. Deleted "to make sure..."